### PR TITLE
os/bluestore: detect unnecessary zeroes in bufferlist when data is written

### PR DESCRIPTION
**Pull request by ljflores**
Bluestore's `_do_write()` method handles writing data from bufferlists. Currently, it writes data from bufferlists without checking it first for zeros. The lack zero detection may negatively impact performance.

The `_do_write()` method handles writing both small and large chunks of data with `_do_write_small()` and `_do_write_large()` respectively. 

In `do_write_small()`, we check if a bufferlist is made up of zeros and avoid writing it. A new counter, `zero_block_count_small`, helps us keep track of how many zero blocks have been skipped in `_do_write_small()`.

Similarly, in `do_write_big()`, we check if a bufferlist is made up of zeros and avoid writing it. A new counter, `zero_block_count_big`, helps us keep track of how many zero blocks have been skipped in `_do_write_big()`.

Trello card: https://trello.com/c/qDn0jspb/754-bluestore-zero-block-detection
Signed-off-by: Laura Flores <lflores@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ljflores** [comment]
[bluestore: zero block detection](https://trello.com/c/qDn0jspb/754-bluestore-zero-block-detection)

**Comment by ljflores** [comment]
[bluestore: zero block detection](https://trello.com/c/qDn0jspb/754-bluestore-zero-block-detection)

File: src/os/bluestore/BlueStore.cc
Comment (ljflores) [reply src/os/bluestore/BlueStore.cc:13584]: That's a good point-- thanks!

File: src/os/bluestore/BlueStore.cc
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:13585]: replace 'metadata' with 'data' here and below?

File: src/os/bluestore/BlueStore.cc
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:14276]: I'm pretty skeptical about this implementation's correctness. Let's run through the following example:
1) original bufferlist t is as follows: t = <0: 4K, non-zero>, <1: 4K, zero>, <2: 4K, non-zero>
2) After zero detection loop one gets: non_zero = <0: 4K, non-zero>, <2: 4K, non-zero>
3) Finally single wctx->write (with all the unmodified lengths and offsets)  is invoked - how would it properly dissect the incoming bufferlist and produce the required gap between units 0 and 2? Highly likely it would just write units 0 and 2 continuously hence breaking the user data. The follow block sequence is presumably a result: <0: 4K, non-zero>, <2: 4K, 
non-zero>, <1: 4K, zero>

File: src/os/bluestore/BlueStore.cc
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:13942]: Given the same pattern to perform zero detection and to apply the write 
I would suggest to introduce a wrapper method to avoid code duplication...

**Review by ifed01 — CHANGES_REQUESTED** [review]
I would strongly encourage to add a UT to store_test suite to make sure the new logics works as expected. Additionally it might be helpful to introduce a perf counter (which can be used in these new UT as well) to track the amount of blocks skipped due to zero detection)

File: src/os/bluestore/BlueStore.cc
Comment (ljflores) [reply src/os/bluestore/BlueStore.cc:14276]: Ah, I see what you mean, I think this is one of those cases where I've been staring at the implementation for so long, that I didn't consider this case. What I intended was for the space to be logically extended between <0 : 4k, non_zero> and <2: 4K, non-zero>, but I see now that they are getting written consecutively.

I will work on fixing this today.

File: src/os/bluestore/BlueStore.cc
Comment (ljflores) [reply src/os/bluestore/BlueStore.cc:13942]: Agreed. I want to leave this for now while I work on fixing the fine-grain implementation, but this will be a fix that will for sure happen once the fine-grain is fixed.

I really appreciate the thorough review, @ifed01!

**Comment by ljflores** [comment]
jenkins test api

**Comment by ljflores** [comment]
jenkins test make check

**Comment by ljflores** [comment]
jenkins test make check arm64

**Comment by ljflores** [comment]
@ifed01 @aclamk @benhanokh @rzarzynski

I had a chat with @jdurgin about this PR and explained my hang-ups on the fine-grain implementation in _do_write_big(). We came to the conclusion that it wouldn't make sense to loop through the buffer and issue a bunch of smaller writes, nor is it possible to issue one big write at the end with chunks of the bufferlist removed, which is what I was initially trying to do. Ultimately, Josh thought the current implementation is sufficient, and that the difference between this and a fine-grain implementation would be negligible.

Thoughts?

Also, the last piece I want to include in this PR is a UT in the the store_test suite (test/objectstore/store_test.cc), which I've almost finished writing. However, I'm unsure how to trigger the store_test suite locally since it doesn't seem to be able to be triggered by ctest. Any advice on that would be helpful.

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by ifed01** [comment]
> @ifed01 @aclamk @benhanokh @rzarzynski
> 
> I had a chat with @jdurgin about this PR and explained my hang-ups on the fine-grain implementation in _do_write_big(). We came to the conclusion that it wouldn't make sense to loop through the buffer and issue a bunch of smaller writes, nor is it possible to issue one big write at the end with chunks of the bufferlist removed, which is what I was initially trying to do. Ultimately, Josh thought the current implementation is sufficient, and that the difference between this and a fine-grain implementation would be negligible.
> 
> Thoughts?
> 
> Also, the last piece I want to include in this PR is a UT in the the store_test suite (test/objectstore/store_test.cc), which I've almost finished writing. However, I'm unsure how to trigger the store_test suite locally since it doesn't seem to be able to be triggered by ctest. Any advice on that would be helpful.

you can just run the relevant binary providing some additional switches, e.g. my script to run it looks like the following:
#!/bin/sh
rm -rf bluestore.test_temp_dir
rm c
export CEPH_ARGS="--plugin_dir=./lib --log-file c --no-log-to-stderr --debug-bluestore 5/10 --debug-bluefs 5/10 --debug-rocksdb 2"
#run everything against bluestore
bin/ceph_test_objectstore --gtest_filter=*/2 --gtest_break_on_failure
#finer filtering 
#bin/ceph_test_objectstore -gtest_filter=ObjectStore/StoreTest.OmapUpgradeRecoveryKeyMatchingTest/2

**Comment by ifed01** [comment]
> @ifed01 @aclamk @benhanokh @rzarzynski
> 
> I had a chat with @jdurgin about this PR and explained my hang-ups on the fine-grain implementation in _do_write_big(). We came to the conclusion that it wouldn't make sense to loop through the buffer and issue a bunch of smaller writes, nor is it possible to issue one big write at the end with chunks of the bufferlist removed, which is what I was initially trying to do. Ultimately, Josh thought the current implementation is sufficient, and that the difference between this and a fine-grain implementation would be negligible.
> 
> Thoughts?

Sounds good to me.

**Comment by ifed01** [comment]
@ljflores - could you please squash the commits and leave complete one(s) only. E.g. no need to implement fine-grain zero detection in one commit and then wipe it out in the following one. IMO for this patch 1-2 commits should be sufficient: 
1. implement zero detection for small writes
2. implement zero detection for big writes
[3. test cases for zero detection]

**Comment by ljflores** [comment]
> rm -rf bluestore.test_temp_dir

Thank you Igor, the `rm -rf bluestore.test_temp_dir` step was what I was missing. I might update the docs or the README to clarify some of these steps for future developers.

**Comment by ljflores** [comment]
> @ljflores - could you please squash the commits and leave complete one(s) only. E.g. no need to implement fine-grain zero detection in one commit and then wipe it out in the following one. IMO for this patch 1-2 commits should be sufficient:
> 
>     1. implement zero detection for small writes
> 
>     2. implement zero detection for big writes
>        [3. test cases for zero detection]

Yes, absolutely. I only left it this way in case anyone wanted to see prior commits. I will squash once I get the unittest in, which I'm almost done writing/testing.

**Comment by ljflores** [comment]
Unit tests have been pushed, and commits have been restructured.

Something I noticed when writing the unit tests is that a couple of counters might need to be included in `_detect_zeros_and_write_small()` and `_detect_zeros_and_write_big()` that are not currently included:
```
l_bluestore_write_small
l_bluestore_write_small_bytes

l_bluestore_write_big
l_bluestore_write_big_bytes
```
Including these would just be a matter of moving them into the appropriate methods and slightly adjusting the unit tests. The main question I have is: Should these counters only be incremented if a bufferlist is actually written (move them to `_detect_zeros_and_write_small/big()`, or do we want them to increment regardless (keep them where they are at the top of `do_write_small/big()`?

**Comment by ljflores** [comment]
jenkins test make check arm64

File: src/os/bluestore/BlueStore.cc
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:14751]: it might be helpful to log relevant notification on bypassing zero block at dout(20) . The same applies to do_write_small

File: src/os/bluestore/BlueStore.cc
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:14232]: as you're changing this piece of code I would suggest to uniform this  logging with _do_write_big and log some notice about writing unconditionally. And similarly indicate relevant reuse_blob value.

File: src/test/objectstore/store_test.cc
Comment (ifed01) [inline src/test/objectstore/store_test.cc:7241]: IMO the following testing scenarios (for both big and small writes) should be extended with 
1. read-after-zero-write checking, i.e. read from the object and make sure zeros are there
2. read-after-zero-overwrite case, i.e. write non-zero first, then overwrite (may be partially?) with zeros and check the content is as expected

File: src/test/objectstore/store_test.cc
Comment (ifed01) [inline src/test/objectstore/store_test.cc:7244]: IIUC this patch is irrelevant to smr drive writing as long as test case performs append only. So may be it makes sense to apply this testing to smr drives - at least partially.

File: src/test/objectstore/store_test.cc
Comment (ljflores) [reply src/test/objectstore/store_test.cc:7241]: Done! Wrote this in a separate commit in case you want to compare to the original, but I will squash once approved.

File: src/test/objectstore/store_test.cc
Comment (ljflores) [inline src/test/objectstore/store_test.cc:7276]: I was going to refer to the "expected" bufferlist as `exp` to follow the pattern that's seen in other parts of the code, but it seems that `exp` is a designated keyword in C++. Thus, I went with `_exp`.

This should probably be fixed in a different PR.

File: src/os/bluestore/BlueStore.cc
Comment (ljflores) [reply src/os/bluestore/BlueStore.cc]: I addressed this comment, but see if you agree with the change in the most recent commit.
I realized that the "reuse_blob" parameter I had added actually communicates the same information as the existing "_new_blob" parameter (if _new_blob is true, a new blob is used; if _new_blob is false, an existing blob is reused). I updated the log message to print unconditionally, but using the "_new_blob" parameter instead of "reuse_blob".

File: src/os/bluestore/BlueStore.cc
Comment (aclamk) [reply src/os/bluestore/BlueStore.cc]: > Given the same pattern to perform zero detection and to apply the write
> I would suggest to introduce a wrapper method to avoid code duplication...
@ifed01 
I think that once we got rid of fine-grain zero detection having separate wrappers is more harm then good.
It passes a ton of parameters and provides limited usability.
@ljflores But ultimately, this is your decision to take.

Comment (ljflores) [reply src/os/bluestore/BlueStore.cc]: > It passes a ton of parameters and provides limited usability.

I see what you mean. The new functions do pass a lot of parameters, which I understand can be confusing.

On the flip-side, it does help to ensure consistency between zero checks and writes, especially in `_do_write_small()` when it happens so often.

Before I go refactoring that part, since it's more of a structural issue, I would like to have the logic and unit tests approved. But perhaps at the end, if we decide that the logic and unit tests look good, but the extra functions are too messy, I can make a last change to refactor the code and write the checks how they were initially. I'm also curious to see what @ifed01 thinks.

**Comment by ljflores** [comment]
Rebased to include the fix for "readthedocs".

**Comment by ljflores** [comment]
jenkins test make check

**Comment by ljflores** [comment]
jenkins test api

**Comment by ljflores** [comment]
jenkins test api

File: src/test/objectstore/store_test.cc
Comment (ifed01) [inline src/test/objectstore/store_test.cc:7313]: Sorry I meant a bit different overwrite scenario. 
Appending write means that  writing extends object size and written block doesn't invalidate any existing data.
Contrary overwrite one doesn't change object size and invalidates  existing block. 
So at [1] you append zeros and at [3] overwrite data with non-zero block. What's missing is overwriting non-zero block with zero one.

Some more versatile write patterns would be nice to have too, e.g. mixing small/big and zero/non-zero writes (e.g. overwriting 'big' non zero block with small zero one or vice verso),  aligned/unaligned small zero blocks, big zero blocks longer than 4K  etc... But that's rather optional, up to you to add that...

And please note that overwrites (as per above specification) aren't currently supported by smr drives, appends  only - hence that is_smr check isn't absolutely useless. You might want either selectively bypass overwrite subcases  or have simple append-only test case(s) supporting SMR and more versatile ones which are omitted in case of SMRs.

Comment (ljflores) [reply src/test/objectstore/store_test.cc:7313]: @ifed01 I see what you mean here-- I misunderstood the original prompt. I've been working on some new changes that I'll push here, and I'll re-request your review when I'm sure all the tests are passing.

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by ljflores** [comment]
jenkins test api

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by ljflores** [comment]
@ifed01 @aclamk 

Fixed the unit tests per @ifed01's specifications. I divided them up into 4 tests `ZeroBlockDetection{Small/Big}{Append/Overwrite}` to make differentiating between cases easier. Let me know if the append cases and overwrite cases make sense now. If they do, I feel confident in the amount of test scenarios for this patch.

Another thing I did was implement two more counters (`zero_block_bytes{small/big}`). I implemented these counters to help us keep track of how many bytes are skipped during zero block detection. This way, it won't be necessary to move the `write_{small/big}` and `write_{small/big}_bytes` counters around, as we will be able to compare those counters against the new zero block counters. I went with this design to preserve as much of the original code implementation as possible.

On that note, I also removed the wrapper functions per @aclamk's comment (https://github.com/ceph/ceph/pull/43337#discussion_r764946021), as I agree with him that it creates too many parameters. The code duplication isn't that much, and I think this helps to prevent making unnecessary changes to the existing code. Also, I believe it's more readable this way, and easier for future developers to customize the dout messages they want for individual cases.

Finally, I ran this branch through Teuthology. I linked the initial run, as well as a re-run I did of failed/dead jobs. Failures are summarized below. I feel confident that this branch is not breaking anything significant.

------------------------------------------------------------------------------------------------------------------------------------------------------------------------

http://pulpito.front.sepia.ceph.com/lflores-2021-12-19_03:36:08-rados-wip-bluestore-zero-detection-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/lflores-2021-12-19_18:26:29-rados-wip-bluestore-zero-detection-distro-default-smithi/

Failures, unrelated:
    6572638 -- timeout expired in wait_until_healthy -- https://tracker.ceph.com/issues/53448
    6572650, 6572644 -- failed in master: http://pulpito.front.sepia.ceph.com/yuriw-2021-12-18_18:14:24-rados-wip-yuriw-master-12.18.21-distro-default-smithi/6569986/
    6572643, 6572648 -- https://tracker.ceph.com/issues/53499

Details:
    Bug_#53448: cephadm: agent failures double reported by two health checks - Ceph - Orchestrator
    Bug_#53499: test_dashboard_e2e.sh Failure: orchestrator/02-hosts-inventory.e2e failed. - Ceph - Mgr - Dashboard

File: src/test/objectstore/store_test.cc
Comment (ifed01) [inline src/test/objectstore/store_test.cc:7281]: nit: overkill IMO, in.is_zero does the trick

File: src/test/objectstore/store_test.cc
Comment (ifed01) [inline src/test/objectstore/store_test.cc:7303]: overkill, following comparison against _exp is sufficient

**Review by ifed01 — APPROVED** [review]
generally LGTM, just some nits which are up to @ljflores  to fix or not.

**Comment by ljflores** [comment]
> generally LGTM, just some nits which are up to @ljflores to fix or not.

Those suggestions make sense-- I can fix them. Thanks for the careful reviews, @ifed01!

File: src/os/bluestore/BlueStore.cc
Comment (ljflores) [reply src/os/bluestore/BlueStore.cc:14281]: I want to keep `b_off` and `b_off0` where they are so I can use them in the "(bl.is_zero())" dout message, but I will move `BlobRef b` and others like that.

File: src/test/objectstore/store_test.cc
Comment (ljflores) [reply src/test/objectstore/store_test.cc]: Done

File: src/test/objectstore/store_test.cc
Comment (ljflores) [reply src/test/objectstore/store_test.cc]: Fixed

**Comment by ljflores** [comment]
jenkins test make check arm64

**Comment by ljflores** [comment]
@aclamk let me know if you have any final comments before I mark this as "needs-qa".

File: src/os/bluestore/BlueStore.cc
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:14293]: tip: it would be nicer if brackets '()' in comment were balanced

File: src/os/bluestore/BlueStore.cc
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:14294]: std::hex missing.

File: src/os/bluestore/BlueStore.cc
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:14540]: std::hex

File: src/os/bluestore/BlueStore.cc
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:14590]: One pair of brackets can be removed.

File: src/os/bluestore/BlueStore.cc
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:14603]: std::hex

**Review by aclamk — CHANGES_REQUESTED** [review]
Good work. Minor details in douts and comments.

**Comment by ljflores** [comment]
Failures, unrelated:

6580187, 6580436 -- https://tracker.ceph.com/issues/52124
6580226, 6580440 -- https://tracker.ceph.com/issues/38455
6580242-- https://tracker.ceph.com/issues/53499
6580330 -- https://tracker.ceph.com/issues/53681
6580439 -- https://tracker.ceph.com/issues/53723
6580078, 6580296 -- https://tracker.ceph.com/issues/53424
6580192 -- https://tracker.ceph.com/issues/51076

Details:

Bug_#52124: Invalid read of size 8 in handle_recovery_delete() - Ceph - RADOS
Bug_#38455: Test failure: test_module_commands (tasks.mgr.test_module_selftest.TestModuleSelftest): RuntimeError: Synthetic exception in serve - Ceph - Mgr
Bug_#53499: test_dashboard_e2e.sh Failure: orchestrator/02-hosts-inventory.e2e failed. - Ceph - Mgr - Dashboard
Bug_#53681: Failed to extract uid/gid for path /var/lib/ceph - Ceph - Orchestrator
Bug_#53723: Cephadm agent fails to report and causes a health timeout - Ceph - Orchestrator
Bug_#53424: CEPHADM_DAEMON_PLACE_FAIL in orch:cephadm/mgr-nfs-upgrade/ - Ceph - Orchestrator
Bug_#51076: "wait_for_recovery: failed before timeout expired" during thrashosd test with EC backend. - Ceph - RADOS

**Comment by yuriw** [comment]
this was tested ref: https://trello.com/c/im0LpXz9

**Comment by ljflores** [comment]
jenkins test api

**Review by aclamk — APPROVED** [review]

**Comment by ljflores** [comment]
Will merge this since it has been tested/approved.

**Comment by idryomov** [comment]
This broke `rbd create --thick-provision` and in general quite a big semantic change.  Punching holes is definitely not what was proposed in https://bugzilla.redhat.com/show_bug.cgi?id=1924129 -- we envisioned a moral equivalent of `fallocate` with `mode=FALLOC_FL_ZERO_RANGE`.  I filed https://tracker.ceph.com/issues/55521 to track this regression.

**Comment by ljflores** [comment]
> This broke `rbd create --thick-provision` and in general quite a big semantic change. Punching holes is definitely not what was proposed in https://bugzilla.redhat.com/show_bug.cgi?id=1924129 -- we envisioned a moral equivalent of `fallocate` with `mode=FALLOC_FL_ZERO_RANGE`. I filed https://tracker.ceph.com/issues/55521 to track this regression.

Thanks for bringing this up, @idryomov. I'll have a look.

**Pull request merged**
